### PR TITLE
Build web

### PR DESCRIPTION
**Pull request by Sybrand**
- Changed API build config to use best effort. (Specifying 8 -> 16GB eats up our entire quote for the tools project)
- Re-factored a lot of the pipeline to "make more sense". (Since we now have api, web + cronjob)
- Re-named some variables that were bothering me.
- [x] Front end builds by itself in it's own build config.
- [x] Front end builds as part of our pipeline
- [x] Front end build + image stream is cleaned up when PR is closed / merged.
- [x] Front end image is tagged and promoted correctly on deployment.
# Test Links:
[Percentile Calculator](https://wps-pr-2050.apps.silver.devops.gov.bc.ca/)
[MoreCast](https://wps-pr-2050.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-2050.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-2050.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-2050.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Fire Behaviour Advisory](https://wps-pr-2050.apps.silver.devops.gov.bc.ca/fire-behaviour-advisory)
[HFI Calculator](https://wps-pr-2050.apps.silver.devops.gov.bc.ca/hfi-calculator)
[FWI Calculator](https://wps-pr-2050.apps.silver.devops.gov.bc.ca/fwi-calculator)

### CODE REVIEW

File: Dockerfile.web [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+RUN npm i yarn && yarn install --frozen-lockfile
+# RUN yarn run build
+
+# Switch back to default user
+USER 1001
+
+# while testing stuff out:
+# docker build -f ./Dockerfile.web --tag=wps/web .
+# -p 8080:3000 - "map port 3000 in the container, to 8080 in the host"
+# docker run -p 3000:3000 wps/web
```
Comment (conbrad) [inline Dockerfile.web:21]: Is this intended to be commented out?

**Review by conbrad — CHANGES_REQUESTED** [review]
Can we remove the web build step from build image, if we have build web image now?
https://github.com/bcgov/wps/runs/6481169395?check_suite_focus=true#step:4:151

**Comment by andrea-williams** [comment]
Any particular reason why this is still a draft PR if it's ready for review?

File: .github/workflows/deployment.yml [hunk 2]
```diff
@@ -74,9 +74,27 @@ jobs:
...
+      - name: Build Image
+        shell: bash
+        run: |
+          oc login "${{ secrets.OPENSHIFT_CLUSTER }}" --token="${{ secrets.OC4_TOOL_TOKEN }}"
+          GIT_BRANCH=${GITHUB_HEAD_REF} NAME_APP=wps-web DOCKER_FILE=Dockerfile.web PATH_BC=openshift/templates/build.web.bc.yaml bash openshift/scripts/oc_build.sh ${SUFFIX} apply
+
   deploy-dev:
     name: Deploy to Dev
-    needs: [build-image, prepare-dev-database]
+    needs: [build-image, build-web-image, prepare-dev-database]
```
Comment (andrea-williams) [inline .github/workflows/deployment.yml:97]: I presume `build-image` refers to the backend image? And then `deploy-dev` means deploying both back- & front-end images, is that correct?

Comment (Sybrand) [reply .github/workflows/deployment.yml:97]: Good point. I'll rename "build-image" to "build-api-image". This story doesn't look at deploying yet - so I'll leave "deploy-dev" alone for now.

File: openshift/templates/build.bc.yaml [hunk 3]
```diff
@@ -64,10 +64,10 @@ objects:
       resources:
         limits:
           cpu: "0"
-          memory: 16Gi
+          memory: "0"
```
[the remark this answers is not in the document]
Comment (conbrad) [reply openshift/templates/build.bc.yaml:67]: `react-scripts` has got to go...

File: Dockerfile.web (hunk 1 above)
Comment (Sybrand) [reply Dockerfile.web:21]: Yes - we don't have the actual build in place - so this is very temporary just helping me. This entire file is a placeholder.

**Review by andrea-williams — APPROVED** [review]

File: Dockerfile.web [hunk 4]
```diff
@@ -9,7 +9,7 @@ FROM registry.access.redhat.com/ubi8/nodejs-16 as static
 USER 0
 
 ADD web .
-RUN npm i yarn && yarn install --frozen-lockfile
+RUN npm i yarn && NODE_OPTIONS=--max_old_space_size=1000 yarn install --frozen-lockfile
```
Comment (conbrad) [inline Dockerfile.web:12]: Can we round this to 1024 bytes? Also some sizing suggestions: https://stackoverflow.com/a/63495296

**Comment by Sybrand** [comment]
Openshift resource issues - closing for now...

File: .vscode/settings.json [hunk 5]
```diff
@@ -28,6 +28,9 @@
         "typescriptreact",
     ],
     "eslint.run": "onSave",
+    "editor.rulers": [
```
Comment (conbrad) [inline .vscode/settings.json:31]: What does these changes do?

File: openshift/scripts/oc_deploy.sh [hunk 6]
```diff
@@ -15,7 +15,7 @@ source "$(dirname ${0})/common/common"
 #%
```
Comment (conbrad) [inline openshift/scripts/oc_deploy.sh:15]: Why did the ACL change for this file?
100755 → 100644

**Review by conbrad — APPROVED** [review]
I like the cleanup thanks. Approved assuming it works (how to test?!)

File: .vscode/settings.json (hunk 5 above)
Comment (Sybrand) [reply .vscode/settings.json:31]: I find having a ruler for 100 very useful in vscode. It doesn't enforce anything, but when doing code reviews on the github user interface, 100 characters is where it starts wrapping lines. So anything below 100 lines is very neat when doing code reviews.

I try to keep my lines to 100 for this reason - and really want a ruler. I can move it to my personal settings if it's an issue.

File: openshift/scripts/oc_deploy.sh (hunk 6 above)
Comment (Sybrand) [reply openshift/scripts/oc_deploy.sh:15]: Good catch - I was moving the file around and it must have changed! No worries, I'll `chmod +x` for the next push.

**Comment by Sybrand** [comment]
> I like the cleanup thanks. Approved assuming it works (how to test?!)

I'm going to close/re-open the PR a few times to make sure there aren't any straggling bit's that aren't being cleaned up, and to make sure that all the cronjobs etc. are being created ok.

File: .vscode/settings.json (hunk 5 above)
Comment (conbrad) [reply .vscode/settings.json:31]: I'm fine with it, just wasn't expecting it. There's just a trailing comma on the `"editor.formatOnSave": true,` setting below.

Comment (Sybrand) [reply .vscode/settings.json:31]: I love trailing commas. It means when you add another line, the unchanged line doesn't come up as part of the code review.

Comment (conbrad) [reply .vscode/settings.json:31]: I know..... it's just a visual bug for me so ignore me :)

**Review by andrea-williams — APPROVED** [review]

**Pull request merged**